### PR TITLE
Set CacheControl for static resources from properties file

### PR DESCRIPTION
**Pull request by tinexw**
First attempt at providing a solution for #9432 

Note that the properties for ``public`` and ``private`` are currently called ``cache-public`` and ``cache-private``. I'm not sure if it's possible to change that? I can look into it if the overall approach is ok.

### CODE REVIEW

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ResourceProperties.java [hunk 1]
```diff
@@ -89,6 +112,36 @@ public Chain getChain() {
...
+			return CacheControl.createCacheControl(this.cacheControl);
+		}
+		return null;
+	}
+
+	@PostConstruct
+	public void checkIncompatibleCacheOptions() {
+		Assert.state(this.cachePeriod == null || this.cacheControl == null,
+				"Only one of cache-period or cache-control may be set.");
+		if (this.cacheControl != null) {
```
[the remark this answers is not in the document]
Comment (tinexw) [reply spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ResourceProperties.java]: It's asserted that not both ``cachePeriod`` and ``cacheControl``are set. So ``cacheControl`` can be null at this point.

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ResourceProperties.java [hunk 2]
```diff
@@ -266,4 +320,173 @@ public void setVersion(String version) {
...
+
+		public Long getsMaxAge() {
+			return this.sMaxAge;
+		}
+
+		public void setsMaxAge(Long sMaxAge) {
+			this.sMaxAge = sMaxAge;
+		}
+
+		static org.springframework.http.CacheControl createCacheControl(CacheControl cacheControlProperties) {
```
[the remark this answers is not in the document]
Comment (tinexw) [reply spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ResourceProperties.java]: > it looks weird for me [...]

Yes, it looks weird to me too... But the ``org.springframework.http.CacheControl``doesn't have a public constructor or any setters. It it itself more like a builder (all methods return the instance) that builds a string (the cache header). 

> Also "createCacheControl" inside a class named "CacheControl" [...] 

Done and I also renamed the CacheControl class to CacheControlProperties

> the method might be an instance method 

Changed

File: spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ResourceProperties.java [hunk 3]
```diff
@@ -266,4 +320,173 @@ public void setVersion(String version) {
...
+			if (cacheControlProperties.getStaleIfError() != null) {
+				cacheControl = cacheControl.staleIfError(
+						cacheControlProperties.getStaleIfError(),
+						TimeUnit.SECONDS);
+			}
+			if (cacheControlProperties.getsMaxAge() != null) {
+				cacheControl = cacheControl.sMaxAge(
+						cacheControlProperties.getsMaxAge(),
+						TimeUnit.SECONDS);
+			}
```
[the remark this answers is not in the document]
Comment (tinexw) [reply spring-boot-autoconfigure/src/main/java/org/springframework/boot/autoconfigure/web/ResourceProperties.java]: I extracted some parts but not sure if it really is better know.

File: spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/web/servlet/WebMvcAutoConfigurationTests.java [hunk 4]
```diff
@@ -797,6 +803,72 @@ public void httpMessageConverterThatUsesConversionServiceDoesNotCreateACycle() {
 				.run((context) -> assertThat(context).hasNotFailed());
 	}
 
+	@Test
+	public void cachePeriod() throws Exception {
+		this.contextRunner.withPropertyValues("spring.resources.cache-period:5")
+				.run((context) -> {
```
[the remark this answers is not in the document]
Comment (tinexw) [reply spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/web/servlet/WebMvcAutoConfigurationTests.java]: Done

File: spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/web/servlet/WebMvcAutoConfigurationTests.java [hunk 5]
```diff
@@ -797,6 +803,72 @@ public void httpMessageConverterThatUsesConversionServiceDoesNotCreateACycle() {
...
+				});
+
+	}
+
+	@Test
+	public void cacheControl() throws Exception {
+		this.contextRunner
+				.withPropertyValues("spring.resources.cache-control.max-age:5",
+						"spring.resources.cache-control.proxy-revalidate:true")
+				.run((context) -> {
```
[the remark this answers is not in the document]
Comment (tinexw) [reply spring-boot-autoconfigure/src/test/java/org/springframework/boot/autoconfigure/web/servlet/WebMvcAutoConfigurationTests.java]: Done

**Comment by tinexw** [comment]
@bclozel I addressed your comments, could you have another look, please?

**Comment by bclozel** [comment]
@tinexw I'll definitely polish (if needed) and merge this - it's scheduled for the next milestone. For the record, @fmcejudo is a contributor, just like you.

Thanks for your contribution, really appreciated!

**Comment by tinexw** [comment]
@bclozel Sorry, I didn't realize it wasn't you. @fmcejudo I addressed your comments, could you have another look, please?

**Comment by bclozel** [comment]
Hi @tinexw - I've merged your changes in e2bc90b and polished things a bit on top. Nothing major, just stuff that happened recently:

* I've changed my mind about `spring.resources.cache-period` and `spring.resources.cache-control` excluding each other; Spring MVC is using `CacheControl` if both are set, so I chose to mirror that behaviour.
* Since #11080, we now use the `Duration` type for such configuration keys.

Thanks again for this new contribution!

**Comment by tinexw** [comment]
@bclozel Thanks!

**Pull request closed without merging**